templates/commands/render.go

```diff
@@ -141,25 +161,155 @@ func (r *Render) parseFlags(args []string) error {
 
-	if err := destOK(useFS, r.flagDest); err != nil {
+	if err := destOK(fSys, r.flagDest); err != nil {
 		return err
 	}
 
-	return fmt.Errorf("not implemented")
+	gg := r.testGetter
+	if gg == nil {
+		gg = &getter.Client{
+			Getters:       getter.Getters,
+			Decompressors: getter.Decompressors,
+		}
+	}
+
+	if r.logger == nil {
+		r.logger = logging.NewFromEnv("ABC_")
+	}
+
+	wd, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("os.Getwd(): %w", err)
+	}
+
+	return r.realRun(ctx, &runParams{
+		fs:           fSys,
+		getter:       gg,
+		cwd:          wd,
+		tempDirNamer: tempDirName,
+	})
+}
+
+type runParams struct {
+	fs     renderFS
+	getter getterClient
+	cwd    string
+
+	// Constructs a name for a temp directory that doesn't exist yet and won't
+	// collide with other directories. Doesn't actually create a directory, the
+	// caller does that. This accommodates quirky behavior of go-getter that
+	// doesn't want the destination directory to exist already.
+	tempDirNamer func(debugID string) (string, error)
+}
+
+// realRun is for testability; it's Run() with fakeable interfaces.
+func (r *Render) realRun(ctx context.Context, rp *runParams) (outErr error) {
+	templateDir, err := r.copyTemplate(ctx, rp)
+	if templateDir != "" { // templateDir might be set even if there's an error
+		defer func() {
+			outErr = errors.Join(outErr, r.maybeRemoveTempDirs(rp.fs, templateDir))
+		}()
+	}
+	if err != nil {
+		return err
+	}
+
+	_ = templateDir // TODO: add template rendering logic
+
+	return nil
+}
+
+// Downloads the template and returns the name of the temp directory where it
+// was saved. If error is returned, then the returned directory name may or may
+// not exist, and may or may not be empty.
+func (r *Render) copyTemplate(ctx context.Context, rp *runParams) (string, error) {
+	templateDir, err := rp.tempDirNamer("template-copy")
+	if err != nil {
+		return "", err
+	}
+	req := &getter.Request{
+		Src:     r.source,
+		Dst:     templateDir,
+		Pwd:     rp.cwd,
+		GetMode: getter.ModeAny,
+	}
+
+	res, err := rp.getter.Get(ctx, req)
+	if err != nil {
+		return templateDir, fmt.Errorf("go-getter.Get(): %w", err)
+	}
+
+	logger.Debugf("copied source template %q into temporary directory %q", r.source, res.Dst)
+	return templateDir, nil
+}
+
+// Calls RemoveAll on each temp directory. A nonexistent directory is not an error.
+func (r *Render) maybeRemoveTempDirs(fs renderFS, tempDirs ...string) error {
+	if r.flagKeepTempDirs {
+		logger.Infof("keeping temporary directories due to --keep-temp-dirs. Locations are: %v", tempDirs)
+		return nil
+	}
+	logger.Debugf("removing temporary directories (skip this with --keep-temp-dirs)")
+
+	var merr error
+	for _, p := range tempDirs {
+		merr = errors.Join(merr, fs.RemoveAll(p))
+	}
+	return merr
+}
+
+// Generate the name for a temporary directory, without creating it. debugID is
+// an optional name that can be included to help understand which directory is
+// which during debugging.
```

Comment:
Not sure I understand the use of debug id. It seems it's a fixed string in real run and not used in test?

Reply:
I renamed the variable and rewrote the comment to make this more clear, PTAL.